### ISSUE TITLE
transport: is not respected on seldondeployment 

### ISSUE THREAD

**Issue by omerfsen** [SeldonIO/seldon-core#2540]
## Describe the bug

For another issue i was focusing i needed to create only Rest Endpoint (via transport: ) but it seems it is not respected with seldon-core-operator 1.3.0 (also tried with 1.2.2) 

## To reproduce

1. Install seldon-core-operator 1.3.0
2. Do a sample seldondeployment with following and add **.spec.transport: rest**

```yaml
apiVersion: machinelearning.seldon.io/v1
kind: SeldonDeployment
metadata:
  name: test
  namespace: prod
spec:
  name: test
  transport: rest
  predictors:
  - componentSpecs:
    - spec:
        containers:
        - env:
          - name: TRACING
            value: '1'
          - name: JAEGER_AGENT_HOST
            valueFrom:
              fieldRef:
                fieldPath: status.hostIP
          - name: JAEGER_AGENT_PORT
            value: '5775'
          - name: JAEGER_SAMPLER_TYPE
            value: const
          - name: JAEGER_SAMPLER_PARAM
            value: '1'
          - name: SELDON_LOG_LEVEL #This is for request/response logging
            value: DEBUG
          image: seldonio/mock_classifier_rest:1.3.0
          name: model1
        terminationGracePeriodSeconds: 1
    graph:
      children: []
      endpoint:
        type: REST
      name: model1
      type: MODEL
      logger: 
        url: http://seldon-request-logger.seldon-logs
        mode: all
    name: tracing
    replicas: 1
    svcOrchSpec:
      env:
      - name: TRACING
        value: '1'
      - name: JAEGER_SAMPLER_TYPE
        value: const
      - name: JAEGER_SAMPLER_PARAM
        value: '1'
      - name: JAEGER_AGENT_HOST
        valueFrom:
          fieldRef:
            fieldPath: status.hostIP
      - name: JAEGER_AGENT_PORT
        value: '5775'
      - name: SELDON_LOG_LEVEL  #This is for request/response logging
        value: DEBUG 
 
````

see both grpc and rest VirtualService is created

```
 kubectl get virtualservices.networking.istio.io -n prod|grep test
test-grpc      [istio-system/seldon-gateway]   [*]     26m
test-http      [istio-system/seldon-gateway]   [*]     26m
```

Also

 **kubectl get seldondeployments.machinelearning.seldon.io -n prod test -o yaml**

```yaml

apiVersion: machinelearning.seldon.io/v1
kind: SeldonDeployment
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"machinelearning.seldon.io/v1","kind":"SeldonDeployment","metadata":{"annotations":{},"name":"test","namespace":"prod"},"spec":{"name":"test","predictors":[{"componentSpecs":[{"spec":{"containers":[{"env":[{"name":"TRACING","value":"1"},{"name":"JAEGER_AGENT_HOST","valueFrom":{"fieldRef":{"fieldPath":"status.hostIP"}}},{"name":"JAEGER_AGENT_PORT","value":"5775"},{"name":"JAEGER_SAMPLER_TYPE","value":"const"},{"name":"JAEGER_SAMPLER_PARAM","value":"1"},{"name":"SELDON_LOG_LEVEL","value":"DEBUG"}],"image":"seldonio/mock_classifier_rest:1.3.0","name":"model1"}],"terminationGracePeriodSeconds":1}}],"graph":{"children":[],"endpoint":{"type":"REST"},"logger":{"mode":"all","url":"http://seldon-request-logger.seldon-logs"},"name":"model1","type":"MODEL"},"name":"tracing","replicas":1,"svcOrchSpec":{"env":[{"name":"TRACING","value":"1"},{"name":"JAEGER_SAMPLER_TYPE","value":"const"},{"name":"JAEGER_SAMPLER_PARAM","value":"1"},{"name":"JAEGER_AGENT_HOST","valueFrom":{"fieldRef":{"fieldPath":"status.hostIP"}}},{"name":"JAEGER_AGENT_PORT","value":"5775"},{"name":"SELDON_LOG_LEVEL","value":"DEBUG"}]}}],"transport":"rest"}}
  creationTimestamp: "2020-10-09T10:45:00Z"
  generation: 1
  name: test
  namespace: prod
  resourceVersion: "328648"
  selfLink: /apis/machinelearning.seldon.io/v1/namespaces/prod/seldondeployments/test
  uid: ed33f4f5-3c50-45a0-bf9b-d35e85f38e46
spec:
  name: test
  predictors:
  - componentSpecs:
    - metadata:
        creationTimestamp: null
      spec:
        containers:
        - env:
          - name: TRACING
            value: "1"
          - name: JAEGER_AGENT_HOST
            valueFrom:
              fieldRef:
                fieldPath: status.hostIP
          - name: JAEGER_AGENT_PORT
            value: "5775"
          - name: JAEGER_SAMPLER_TYPE
            value: const
          - name: JAEGER_SAMPLER_PARAM
            value: "1"
          - name: SELDON_LOG_LEVEL
            value: DEBUG
          image: seldonio/mock_classifier_rest:1.3.0
          name: model1
          ports:
          - containerPort: 6000
            name: metrics
            protocol: TCP
          resources: {}
          volumeMounts:
          - mountPath: /etc/podinfo
            name: seldon-podinfo
        terminationGracePeriodSeconds: 1
    engineResources: {}
    graph:
      endpoint:
        service_host: localhost
        service_port: 9000
        type: REST
      implementation: UNKNOWN_IMPLEMENTATION
      logger:
        mode: all
        url: http://seldon-request-logger.seldon-logs
      name: model1
      type: MODEL
    labels:
      version: tracing
    name: tracing
    replicas: 1
    svcOrchSpec:
      env:
      - name: TRACING
        value: "1"
      - name: JAEGER_SAMPLER_TYPE
        value: const
      - name: JAEGER_SAMPLER_PARAM
        value: "1"
      - name: JAEGER_AGENT_HOST
        valueFrom:
          fieldRef:
            fieldPath: status.hostIP
      - name: JAEGER_AGENT_PORT
        value: "5775"
      - name: SELDON_LOG_LEVEL
        value: DEBUG
  transport: rest
status:
  address:
    url: http://test-tracing.prod.svc.cluster.local:8000/api/v1.0/predictions
  deploymentStatus:
    test-tracing-0-model1:
      availableReplicas: 1
      replicas: 1
  replicas: 1
  serviceStatus:
    test-tracing:
      grpcEndpoint: test-tracing.prod:5001
      httpEndpoint: test-tracing.prod:8000
      svcName: test-tracing
    test-tracing-model1:
      httpEndpoint: test-tracing-model1.prod:9000
      svcName: test-tracing-model1
  state: Available

```

as you can see below we have both http and grpc endpoint is created (:5001 and :8000). We expect to only see :8000 is created (namely httpEndpoint)

```yaml
....
      availableReplicas: 1
      replicas: 1
  replicas: 1
  serviceStatus:
    test-tracing:
      grpcEndpoint: test-tracing.prod:5001
      httpEndpoint: test-tracing.prod:8000
      svcName: test-tracing
    test-tracing-model1:
      httpEndpoint: test-tracing-model1.prod:9000
      svcName: test-tracing-model1
  state: Available
...
````

## Expected behaviour

Only http/rest transport should be created so as output i should only get

```
 kubectl get virtualservices.networking.istio.io -n prod|grep test
test-http      [istio-system/seldon-gateway]   [*]     26m
```

## Environment

Seldon Core Operator 1.3.0
Istio: 1.6.8

**Comment by omerfsen** [comment]
This issue is related with https://github.com/istio/istio/issues/27513

**Comment by omerfsen** [comment]
@groszewn I have been told you have some expertise on this and told to link you by @cliveseldon

**Comment by groszewn** [comment]
So the operator definitely is not taking into account the `spec.transport` field when creating services and virtualservices (see [here](https://github.com/SeldonIO/seldon-core/blob/0cfbc38586b387db823aa1be53c044d214173ddf/operator/controllers/seldondeployment_controller.go#L683-L689) and [here](https://github.com/SeldonIO/seldon-core/blob/0cfbc38586b387db823aa1be53c044d214173ddf/operator/controllers/seldondeployment_controller.go#L330-L343), respectively).  

Based on the logic in [createComponents](https://github.com/SeldonIO/seldon-core/blob/0cfbc38586b387db823aa1be53c044d214173ddf/operator/controllers/seldondeployment_controller.go#L373), it seems like `grpcAllowed` is set to `false` (which would result in the grpc virtual service not being created) only when `noEngine` is set.

From what I'm seeing, `spec.transport` is only used to tell the executor container what transport to use (see [here](https://github.com/SeldonIO/seldon-core/blob/0cfbc38586b387db823aa1be53c044d214173ddf/operator/controllers/seldondeployment_engine.go#L253)).

**Comment by axsaucedo** [comment]
@groszewn that is correct, I see what @omerfsen means in regards to both ingress being created, at least so far this is expected - if it's not expected to work on REST/GRPC I could see this being something that we'd want to restrict. @omerfsen is there a particular reason why you need this to be set explicitly by only creating the relevant virtual service?

**Comment by omerfsen** [comment]
@axsaucedo The reason for that was https://github.com/istio/istio/issues/27513 On Seldon Core 1.2.2/12.3 somehow GRPC and REST endpoint is mixing and Istio Pilot is activating GRPC VirtualService instead of Rest Virtual Service so we were getting "upstream error for 30 min until it is fixed automatically" so i wanted to completely disable GRPC endpoint using `transport: rest` but it did not work. 

Although I have tried Seldon Core 1.3.0 and Istio issue (https://github.com/istio/istio/issues/27513) seems to have completely gone but i think this transport issue still stands as somehow some may require specific transport protocol.

**Comment by cliveseldon** [comment]
This will be not relevant when #2378 is fixed in 1.5

**Comment by cliveseldon** [comment]
Closing this as 1.5 is released and REST and gRPC are both exposed for all sdeps.
